iModelCore/ECPresentation/Source/Shared/RulesPreprocessor.cpp

```diff
@@ -978,14 +978,62 @@ bvector<ContentRuleCP> RulesPreprocessor::_GetContentRules()
     return result;
     }
 
+/*---------------------------------------------------------------------------------**//**
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+static bvector<NavNodeKeyCPtr> SplitLabelGroupedKey(IConnectionCR connection, INodeInstanceKeysProvider const& instanceKeyProvider, LabelGroupingNodeKey const& key)
+    {
+    bvector<NavNodeKeyCPtr> result;
+    bmap<ECN::ECClassId, bvector<ECInstanceKey>> instanceKeyMap;
+    instanceKeyProvider.IterateInstanceKeys(key, [&](ECInstanceKey instanceKey)
+        {
+        auto iter = instanceKeyMap.find(instanceKey.GetClassId());
+        if (iter == instanceKeyMap.end())
+            iter = instanceKeyMap.Insert(instanceKey.GetClassId(), {}).first;
+        iter->second.push_back(instanceKey);
+        return true;
+        });
+    for (auto& instanceKeys : instanceKeyMap)
+        {
+        std::unique_ptr<bvector<ECInstanceKey>> keyList = std::make_unique<bvector<ECInstanceKey>>(instanceKeys.second);
+        result.push_back(LabelGroupingNodeKey::Create(connection, key.GetSpecificationIdentifier(),
+            nullptr, key.GetLabel(), instanceKeys.second.size(), nullptr, std::move(keyList)));
+        }
+    return result;
+    }
+
+/*---------------------------------------------------------------------------------**//**
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+static bvector<NavNodeKeyCPtr> GetInputNodeKeys(INavNodeKeysContainerCR inputNodeKeys, IConnectionCR connection, INodeInstanceKeysProvider const* instanceKeyProvider)
+    {
+    bvector<NavNodeKeyCPtr> singleNodeKeys;
+
+    for (NavNodeKeyCPtr const& inputNodeKey : inputNodeKeys)
+        {
+        if (inputNodeKey->AsLabelGroupingNodeKey() && instanceKeyProvider != nullptr)
+            {
+            bvector<NavNodeKeyCPtr> keys = SplitLabelGroupedKey(connection, *instanceKeyProvider, *inputNodeKey->AsLabelGroupingNodeKey());
+            ContainerHelpers::Push(singleNodeKeys, keys);
+            }
+        else
+            {
+            singleNodeKeys.push_back(inputNodeKey);
+            }
+        }
+    return singleNodeKeys;
+    }
+
 /*---------------------------------------------------------------------------------**//**
 * @bsimethod
 +---------------+---------------+---------------+---------------+---------------+------*/
 ContentRuleInputKeysContainer RulesPreprocessor::_GetContentSpecifications(ContentRuleParametersCR params)
     {
     ContentRuleInputKeysContainer specs;
     bset<NavNodeKeyCP> handledNodes;
-    for (NavNodeKeyCPtr const& inputNodeKey : params.GetInputNodeKeys())
+    bvector<NavNodeKeyCPtr> singleNodeKeys = GetInputNodeKeys(params.GetInputNodeKeys(), m_connection, params.GetInstanceKeyProvider());

```

Comment:
I think we need to improve naming here:
- Why `singleNodeKeys`?
- `GetInputNodeKeys` -> `ProcessNodeKeysForContent`?
- `SplitLabelGroupedKey` -> `SplitLabelGroupedKeyByClass`?